### ISSUE TITLE
Pre-Checkout Timeout Error while making a Payment

### ISSUE THREAD

**Issue by sojinsamuel** [yagop/node-telegram-bot-api#1187]
i've built a telegram bot using nodejs that downloads videos from tweet urls. Completed that part. now, Iim struggling to implement a paywall using the `sendInvoice` method in the Telegram API.

https://core.telegram.org/bots/api#sendinvoice

i'm using Stripe for payments in test mode. however, after sending the invoice, when users try to pay, it buffers and times out as in the image.

![Screenshot from 2024-03-31 18-59-16](https://github.com/yagop/node-telegram-bot-api/assets/77185816/b9e3b7e2-4269-41dc-ac87-3843df26bdd2)

Question i asked  on stackoverflow (my version 1 code without using any external libs for the bot):
https://stackoverflow.com/questions/78251916/stripe-pre-checkout-timeout-error-while-making-a-test-payment-from-telegram-bot

Not sure what was wrong there at the first time; i did find this similar problem on stackoverflow:
https://stackoverflow.com/questions/73341054/telegram-sendinvoice-method-is-failing

Which i almost thought i solved it, but i am not even receiving the `update` object as explained here
https://core.telegram.org/bots/api#answershippingquery

So finally to implement what's being said and to make things simple i started to use this library  `node-telegram-bot-api` (previously used rest api from telegram)

Even after making the changes suggested from the docs and on the similar stackoverflow solution above.

I have come to this (an abstracted version 2 of my source code):

```javascript
import { config } from "dotenv";
import TelegramBot from "node-telegram-bot-api";
import getTwitterVideoURL from "./twitterfunction.js";

config();

const token = process.env.TELEGRAM_BOT_TOKEN;
// console.log(token);
const bot = new TelegramBot(token, { polling: true });

bot.on("message", async (msg) => {
  //   console.log(JSON.stringify(msg, null, 2));
  const chatId = msg.chat.id;
  const message = msg.text.trim();

  console.log(JSON.stringify(msg, null, 2));

  bot.sendInvoice(
    chatId,
    "Title goes here",
    "Unlock premium content with a subscription",
    "Invoice for payment",
    "MY_PROVIDER_TOKEN",
    "USD",
    [{ label: "tax", amount: 4000 }],
    {
      is_flexible: false,
      need_phone_number: false,
      need_email: true,
      need_shipping_address: false,
    }
  );

});

bot.on("pre_checkout_query", async (query) => {
  console.log("Received pre-checkout query:", query);

  // Answer the pre-checkout query within 10 seconds more info: https://core.telegram.org/bots/api#answerprecheckoutquery
 
  const preCheckoutQueryAnswer = await bot.answerPreCheckoutQuery(
    query.id,
    true
  );

  console.log("Pre-checkout query answer:", preCheckoutQueryAnswer);
});

bot.on("successful_payment", async (payload) => {
  console.log("Successful payment:", payload);

  const message = `Thank you for your payment of ${
    payload.total_amount / 100
  } ${payload.currency}!`;
  await bot.sendMessage(payload.chat_id, message);
});

bot.on("invoice", (error) => {
  console.log("invoice", error);
});

bot.on("shipping_query", async (query) => {
  console.log("Received shipping query:", query);

  // Don't have any shipping options, provide an empty array
  const shippingOptions = [];

  const shippingQueryAnswer = await bot.answerShippingQuery(
    query.id,
    true, // Set to false if the shipping cannot be processed
    { shipping_options: shippingOptions }
  );

  console.log("Shipping query answer:", shippingQueryAnswer);
});
```
Still the same problem. Tried other payment provider too no luck.

Can anyone help me out please.

**Comment by werniq** [comment]
Hi. Have you solved it?

**Comment by sojinsamuel** [comment]
no, I havent worked on it for a while. will you lmk if you find a solution

**Comment by werniq** [comment]
Yes, it is related to pre_checkout_query. So after client clicks Pay button, some pre checkout query is sent to your bot. You need to handle responding to this query, and everything will be ok.

**Comment by sojinsamuel** [comment]
> Yes, it is related to pre_checkout_query. So after client clicks Pay button, some pre checkout query is sent to your bot. You need to handle responding to this query, and everything will be ok.

```javascript
bot.on("pre_checkout_query", async (query) => {
  console.log("Received pre-checkout query:", query);

  // Answer the pre-checkout query within 10 seconds more info: https://core.telegram.org/bots/api#answerprecheckoutquery
 
  const preCheckoutQueryAnswer = await bot.answerPreCheckoutQuery(
    query.id,
    true
  );

  console.log("Pre-checkout query answer:", preCheckoutQueryAnswer);
});
```
But i didnt get any payload after clicking on pay button, is it possible for share what you would think a possible soultion for this issue or an update. anyways thanks a lot for contributing to the conversation :smiley:

**Comment by werniq** [comment]
in python my handler looks like this

```
@dp.pre_checkout_query()
async def pre_checkout_query_handler(pre_checkout_query: types.pre_checkout_query):
    await bot.answer_pre_checkout_query(pre_checkout_query.id, ok=True)

```

**Comment by skynette** [comment]
> Yes, it is related to pre_checkout_query. So after client clicks Pay button, some pre checkout query is sent to your bot. You need to handle responding to this query, and everything will be ok.

how can i achieve this using api calls, im not using any sdk, i generated a payment link using /createInvoiceLink, but when i try to pay it times out

**Comment by dassuumasarat6-sys** [comment]
Hi

**Comment by dassuumasarat6-sys** [comment]
Hi

**Comment by dassuumasarat6-sys** [comment]
Hi

**Comment by dassuumasarat6-sys** [comment]
Hi

**Comment by dassuumasarat6-sys** [comment]
Hi

**Comment by dassuumasarat6-sys** [comment]
Hi

**Comment by dassuumasarat6-sys** [comment]
Hi

**Comment by yagop** [comment]
For anyone landing here: this is a usage issue, not a library bug. After the user taps **Pay**, Telegram sends a `pre_checkout_query` that your bot **must** answer within 10 seconds, or the payment times out (the spinner in the screenshot):

```js
bot.on('pre_checkout_query', (query) => {
  bot.answerPreCheckoutQuery(query.id, true).catch(console.error);
});
```

If you never receive `pre_checkout_query` at all, the cause is upstream of this library:

- Payments must be **set up in @BotFather** (Payments -> connect a provider), and you must use the matching **provider_token** (the Stripe **TEST** token for test mode).
- Don't run two `getUpdates` consumers on the same token, and make sure `pre_checkout_query` isn't excluded by a custom `allowed_updates`.
- `answerPreCheckoutQuery` must succeed within 10s - do slow validation before answering, or answer `true` first.

The library exposes `sendInvoice` / `createInvoiceLink` / `answerPreCheckoutQuery` / `answerShippingQuery` and emits the `pre_checkout_query` / `shipping_query` / `successful_payment` events, so the flow works once the provider config is correct. Closing as a usage/configuration issue.
